src/main/java/pl/damianszczepanik/jenkins/buildhistorymanager/model/actions/DeleteArtifactsMatchingPatternsAction.java

```diff
@@ -0,0 +1,147 @@
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.jenkinsci.remoting.RoleChecker;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+import hudson.FilePath;
+import hudson.FilePath.FileCallable;
+import hudson.Util;
+import hudson.model.Run;
+import hudson.remoting.VirtualChannel;
+import jenkins.util.VirtualFile;
+
+/**
+ * Deletes the artifacts matching patterns.
+ */
+public class DeleteArtifactsMatchingPatternsAction extends Action {
+    private static final Logger LOG = Logger.getLogger(DeleteArtifactsMatchingPatternsAction.class.getName());
+
+    private String includePatterns;
+    private String excludePatterns;
+
+    @DataBoundConstructor
+    public DeleteArtifactsMatchingPatternsAction() {
+        // Jenkins stapler requires to have public constructor with @DataBoundConstructor
+    }
+
+    public String getIncludePatterns() {
+        return includePatterns;
+    }
+
+    @DataBoundSetter
+    public void setIncludePatterns(String includePatterns) {
+        this.includePatterns = includePatterns;
+    }
+
+    public String getExcludePatterns() {
+        return excludePatterns;
+    }
+
+    @DataBoundSetter
+    public void setExcludePatterns(String excludePatterns) {
+        this.excludePatterns = excludePatterns;
+    }
+
+    static boolean isDirectoryEmpty(Path path) throws IOException {
+        try (DirectoryStream<Path> directoryStream = Files.newDirectoryStream(path)) {
+            return !directoryStream.iterator().hasNext();
+        }
+    }
+
+    // if 'file' is on a different node, this FileCallable will be transferred to that node and executed there.
+    public static final class DeleteFileCallable implements FileCallable<Void> {
+        private static final long serialVersionUID = -7712623306942117400L;
+        private final File archiveRootFile;
+
+        public DeleteFileCallable(File archiveRootFile) {
+            this.archiveRootFile = archiveRootFile;
+        }
+
+        @Override 
+        public Void invoke(File file, VirtualChannel channel) throws IOException {
+            deleteFileOrLogError(file);
+            return null;
+        }
+
+        void deleteEmptyDirectoryAndParent(File directory) throws IOException {
+            if (isDirectoryEmpty(directory.toPath()) && hasValidParent(directory.getParentFile())) {
+                Util.deleteFile(directory);
+                deleteParentDirectories(directory.getParentFile());
+            }
+        }
+
+        boolean hasValidParent(File parent) {
+            return this.archiveRootFile.equals(parent);
+        }
+
+        void deleteParentDirectories(File directory) {
+            File parent = directory;
+
+            while (parent != null && !parent.equals(this.archiveRootFile)) {
```

Comment:
in what cases parent can be null? when taking parent from root directory?

Reply:
It appears safe to remove the `parent !=null` condition from the loop. :)